### PR TITLE
[REF] changes the behavior of helper edit in tours

### PR DESCRIPTION
**Pull request by pipu-odoo**
Before this commit, in rounds, the edit() helper does not directly 
trigger blur events and changes after modifying the value of a field 
(input or textarea). This was a desired behavior but which implies that 
it is then necessary to add another event which then triggers the blur 
& change events, for example: 
- && press Enter 
- && press Tab
- && click body
- && click .where_you_want

However, this causes confusion because it is not what is expected by 
users of the API.  

In this commit, to simplify the API, we trigger the blur and change 
events directly in the edit() helper.

task~3922501

https://github.com/odoo/enterprise/pull/63053

### CODE REVIEW

File: addons/web_tour/static/src/tour_service/tour_utils.js [hunk 1]
```diff
@@ -416,6 +427,41 @@ export class RunningTourActionHelper {
...
+     * That's why we prefer **tab** action for inputs.
+     * @param {Node} node
+     * @returns {"blur"|"enter"|"tab"|null}
+     */
+    _get_edit_confirm_action = (node) => {
+        const tagNode = getTag(node, true);
+        const confirmActions = {
+            input: [
+                ["class", "o_searchview_input", null],
+                ["parentClass", ".o_list_view", "click .o_selected_row"],
```
Comment (aab-odoo) [inline addons/web_tour/static/src/tour_service/tour_utils.js:441]:
```suggestion
                ["parentClass", "o_list_view", "click .o_selected_row"],
```
?

File: addons/web_tour/static/src/tour_service/tour_utils.js [hunk 2]
```diff
@@ -281,7 +288,12 @@ export class RunningTourActionHelper {
     edit(text, selector) {
         const element = this._get_action_element(selector);
         hoot.click(element);
-        hoot.edit(text);
+        const confirm = this._get_edit_confirm_action(element);
+        hoot.edit(text, { confirm });
+        if (!["blur", "enter", "tab", null].includes(confirm)) {
+            const action = confirm.split(" ");
+            hoot[action[0]](action[1]);
+        }
```
Comment (aab-odoo) [inline addons/web_tour/static/src/tour_service/tour_utils.js:296]: would be nice to get rid of that

File: addons/web_tour/static/src/tour_service/tour_utils.js (hunk 1 above)
Comment (aab-odoo) [reply addons/web_tour/static/src/tour_service/tour_utils.js:441]: can't we blur (or null) for editable lists?

File: addons/web_tour/static/src/tour_service/tour_utils.js [hunk 3]
```diff
@@ -416,6 +427,41 @@ export class RunningTourActionHelper {
...
+     * @param {Node} node
+     * @returns {"blur"|"enter"|"tab"|null}
+     */
+    _get_edit_confirm_action = (node) => {
+        const tagNode = getTag(node, true);
+        const confirmActions = {
+            input: [
+                ["class", "o_searchview_input", null],
+                ["parentClass", ".o_list_view", "click .o_selected_row"],
+                ["parentClass", "we-list", "blur"],
```
Comment (aab-odoo) [inline addons/web_tour/static/src/tour_service/tour_utils.js:442]: isn't it the default?

File: addons/web_tour/static/src/tour_service/tour_utils.js [hunk 4]
```diff
@@ -416,6 +427,41 @@ export class RunningTourActionHelper {
...
+                ["id", "o_link_dialog_url_input", null], // website
```
Comment (aab-odoo) [inline addons/web_tour/static/src/tour_service/tour_utils.js:443]: couldn't this one be dealt with user space (so a specific run in the tour itself)?

File: addons/web_tour/static/src/tour_service/tour_utils.js [hunk 5]
```diff
@@ -416,6 +427,41 @@ export class RunningTourActionHelper {
...
+                // When autocomplete, we want to see suggestion before trigger confirmAction
+                ["class", "o-autocomplete--input", null],
+                ["parentClass", ".o-web-studio-field-dynamic-placeholder", null],
```
Comment (aab-odoo) [inline addons/web_tour/static/src/tour_service/tour_utils.js:446]: same here

File: addons/web_tour/static/src/tour_service/tour_utils.js [hunk 6]
```diff
@@ -416,6 +427,41 @@ export class RunningTourActionHelper {
...
+            ],
+            textarea: [
+                ["parentClass", ".o_list_view", "click .o_selected_row"],
+                ["class", "o-mail-Composer-input", null],
```
Comment (aab-odoo) [inline addons/web_tour/static/src/tour_service/tour_utils.js:450]: same here (basically, for everything that isn't from web/)

File: addons/web_tour/static/src/tour_service/tour_utils.js (hunk 1 above)
Comment (pipu-odoo) [reply addons/web_tour/static/src/tour_service/tour_utils.js:441]: image

File: addons/web_tour/static/src/tour_service/tour_utils.js (hunk 3 above)
Comment (pipu-odoo) [reply addons/web_tour/static/src/tour_service/tour_utils.js:442]: No ... tab is the default

**Comment by alexkuhn** [comment]
Sometimes you want to edit while keeping focus / selection. You might also ensure some code works without relying on implicit blur / change triggered, which usually happens from another explicit user interaction like pressing `Enter`, `Tab` / `Esc` or click away, each might do something completely different.

I really don't get the "confusion" from requiring to explicitly state the other interaction to blur / ENTER / etc... This matches exactly the interaction expected by a user when inserting or replacing text in an input. I really think this is a bad idea to make edit do more than it really does, as this reduces the amount of actions possible in test or encourage us to not use these helpers and instead trigger events by hand, which I suppose you understand this is not a good thing.

I get this is good to make helper easier to use, but we should not go too far and make it overly opinionated in its use.

Just to give an idea of a problem with this opinionated logic of `edit`: if I want to add a `@mention` in composer and then navigate in the mention suggestion list. This won't be possible with `edit()` because blur removes the suggestion mention list.

**Comment by rdeodoo** [comment]
If I am not wrong, I don't think what was asked was to make it so `edit` do more than edit.
What was problematic with hoot (which was working fine before) was the following:
```js
   {
        content: "edit smth",
        trigger: 'input.smth',
        run: 'edit blabla',
    }, {
        content: "Click somewhere else",
        trigger: 'body',
    },
```

Following hoot, the click on body was not triggering blur/onchange event etc (something like that?). It was not considering the fact when clicking on the body, that it previously was focusing an input and that since the input focus would be lost, it needed to trigger stuff.

I did not really dig into it at the time, it was mainly @detrouxdev which first reported that kind of regression, and @qsm-odoo who had recent looks into failing tours, they will have more input about it

File: addons/web_tour/static/src/tour_service/tour_utils.js (hunk 1 above)
Comment (aab-odoo) [reply addons/web_tour/static/src/tour_service/tour_utils.js]: Of course, but `parentClass`, so you know it's a class... for `o_searchview_input`, you add the dot yourself

**Comment by qsm-odoo** [comment]
> I really don't get the "confusion" from requiring to explicitly state the other interaction to blur / ENTER / etc...

If I understand what we are talking about that's a big confusion:
Step 1. add text in input
Step 2. click elsewhere
=> It would be *highly* confusing if no blur is triggered unless you added "add text in input and blur then click". And mainly, it would just hide a lot of bugs if that blur was not triggered. How was it working before, we really triggered that blur by hand? 

> In this commit, to simplify the API, we trigger the blur and change events directly in the edit() helper.

But I agree that triggering a blur automatically with the `edit` helper seems equally bad.

Basically we should really have what we had before. All "text" for input should have become "edit", all "text_blur" should have become "edit && blur" and that's it... but I guess it's more complicated than that?

**Comment by oomsveta** [comment]
Hello there 👋

Just for the record, I'm currently facing a similar problem.
Recently, a feature of VoIP broke due to a change in OWL: `t-model.trim="field"` is now in lazy mode by default, i.e. the field is now updated on change instead of on input. No tests could detect the problem because the helper I used (`insertText`) systematically fired the "change" event.

If something breaks, but the test helpers prevent it from being caught, that is a problem. That's why I thought at first that the "change" event shouldn't be triggered by the helper.

That was until I was faced with the opposite problem... I have another feature that is expected to be triggered by the "change" event. But it seems like the change event is never fired for an input that is changed programmatically, or at least that's what I infer from the specs and my own observations:
> change: Fired at controls when **the user** commits a value change (see also the input event) 
> https://html.spec.whatwg.org/multipage/indices.html#event-change

With that in mind, I think we should be able to have both, depending on what's needed in the context of the test. The best solution to me would be to programmatically fire the "change" event only when the normal criteria are met. For example, perhaps the helper could add a "blur" event listener on the form control and fire the "change" event whenever that listener is triggered. But if that's not possible / too complicated, I feel like it should at least be possible to provide an option to the helper to enable/disable the firing of the "change" event.

My two cents 🙂

**Comment by aab-odoo** [comment]
@alexkuhn @rdeodoo @qsm-odoo 

TLDR: there's no universal solution, we want to find the good default behavior for the basic `edit` action, which hopefully fits for most of the usecases, and find a way to deal with the others.

First (even though this PR isn't in draft), there's no plan to merge it as is. But this branch helped listing all corner cases we have in the codebase for which edit + blur isn't working.

When we think about an "edit" action, it's not only just about inserting stuff in an input. For instance in views, the change is only "committed" when the `change` event is triggered. Only then, the value is sent to the model, and potential onchanges are triggered. I think that from the perspective of a guy writing a tour, he doesn't care about framework details. He just wants to say, edit that field, and that's it. So if editing that field is supposed to trigger onchanges, those onchanges should be applied by the next step, without relying on the fact that a next step clicks somewhere, which blurs the input and thus triggers the `change` (I think that's how it works currently).

I also think that we don't want to have to explicitly say, in the step, that the edition is followed by Enter, Tab, blur or whatever. This is too low-level details for a tour, imho (i.e. we don't want to see "edit && Enter" everywhere).

Obviously this logic doesn't apply to everything. The idea was to determine how far we can go with a simple `edit` function that simply triggers "Enter" in the input. If that covers 80-90% of usecases, I think that's fine. Then we should consider how to deal with the remaining. Historically, we had a custom handling for autocomplete inputs. They are widely used, and the feature comes from the framework, so I think it's ok to have a "magic" special case for it (i.e. `edit` won't do the "Enter" if the input is an autocomplete). For the rest, I'm not against more specific actions like `edit_blur` or `edit_tab` if that's a common usecase. Or keeping `text` which only inserts text content, but doesn't trigger a change/blur, so e.g.:
 - `edit abc` => insert "abc" + simulates "Enter" that triggers the `change` event)
 - `text abc` => only insert "abc" (focus remains on the input, no `change` event fired, only `input` events)

A word about why this behaves differently/why we can't really do like in legacy: 
In legacy (both tours and qunit), we manually fired __some__ events (those we needed for the code to react). With hoot, we changed the approach: we now try to mock real user interactions. In other words, before we fired a `change` event. Now, we simulate an `Enter`, which triggers the `change`, and all other events that are fired when you really press `Enter`.

**Comment by qsm-odoo** [comment]
@aab-odoo (sorry for the long message, btw: not trying to be annoying but the tour system is critical for the website builder, we cannot at the moment have anything else easily).

> He just wants to say, edit that field, and that's it.

Exactly: enter something in that field :innocent: 

> So if editing that field is supposed to trigger onchanges, those onchanges should be applied by the next step, without relying on the fact that a next step clicks somewhere, which blurs the input and thus triggers the change

I guess you mean "before the next step", in that case, I don't agree. The tour system has to describe the real user flow. And the real user flow is that onchanges are triggered once you blur, that is, in practice, once you **click on or focus something else**. If you really do want to see what happens on blur in one go, then it has to be explicit (like `text_blur` or `edit && blur`).

> I also think that we don't want to have to explicitly say, in the step, that the edition is followed by Enter, Tab, blur or whatever. This is too low-level details for a tour, imho (i.e. we don't want to see "edit && Enter" everywhere).

I guess that's debatable. For me, tours **are** a very low level thing. Both testing usecases do exist (as advertised in previous messages), I'd say that `edit` + `edit && blur` would be better that `edit` + `edit && noblur` :wink: 

What is not debatable IMO is that "a click made somewhere else" made by a specific step has to trigger any blur/change it would trigger in the real user flow (e.g. if an input was focused). If somehow this was not the case always before, I would be very surprised, I don't see how we could test things properly without that :scream: 

> Obviously this logic doesn't apply to everything. The idea was to determine how far we can go with a simple edit function that simply triggers "Enter" in the input. If that covers 80-90% of usecases, I think that's fine. 

The thing is: it will cover 99% of the use cases **but by mistake**. You are not simulating what a real user would do and IMO, not what your test would describe. To come back to your onchange case:
Step 1: edit field
Step 2: click on save
=> With your implementation you don't test the actual user flow of "changing the value of the field then hitting save", you are testing "changing the value of the field, click somewhere else then hitting save" **which most users don't do** (I do, I click everywhere like a crazy person but that's not the case for everyone) (and IMO: not what the test seems to describe). And that edit + save case was actually a recurring buggy thing.

> For the rest, I'm not against more specific actions like edit_blur or edit_tab if that's a common usecase

You seems to be rather asking for an `edit_no_blur` no? Or I did not understand anything ^^

> Or keeping text which only inserts text content, but doesn't trigger a change/blur, so e.g.:

Well, maybe... that's what we want in the end: two utils for the two usecases. But IMO text and edit are just gonna get confused, why not explicit like before (especially given, again, I think "no blur" should be the default when you write the test without thinking about it).

> we now try to mock real user interactions. In other words, before we fired a change event. Now, we simulate an Enter, which triggers the change

Not sure how it is different from before: we did try to mock real user interactions. Maybe we did it badly though, I believe you :wink: But the case "enter text in step 1 and click in step 2" was working no? Why was it working before by just triggering some events and not now without an explicit `&& blur` (not saying the system was good before, I know it much needed refactoring btw).

**Comment by aab-odoo** [comment]
I agree that it's better to have a `"edit" + "blur"` than `"edit" + "noblur"`, except if 99% of the time bluring is ok.

> To come back to your onchange case:
Step 1: edit field
Step 2: click on save
=> With your implementation you don't test the actual user flow of "changing the value of the field then hitting save"

My point is that isn't the purpose of tours, it's the purpose of unit tests (which are more "low-level"). I don't want to have to write a tour to assert that I can edit an char field and click on save directly without bluring the input first. But we have at least one unit test for this, of course. 

What are the usecases where hitting "Enter" is problematic?

**Comment by aab-odoo** [comment]
And btw the former version triggered a `change` event when editing, so it was basically the same, except that we didn't really triggered the action that produces the `change` event (e.g. blur, Enter).

**Comment by qsm-odoo** [comment]
> And btw the former version triggered a change event when editing, so it was basically the same, except that we didn't really trigger the action that produces the change event (e.g. blur, Enter).

Ok, then I guess it's not worth arguing which is better. If we triggered "change" without "blur" before, then that's the closest of the "edit + Enter" you wanted, let's do that if you think it's the best, that would lead to the least change of behavior in tests :+1: Then no need of blurring (the enter without blur triggers the change) and we can keep the `&& blur` syntax (which then replaces `text_blur` as I wanted). Then later if we have usecase where we don't want the "Enter + change" part (like I thought we wanted), then we can think about it at that moment :shrug: 

Still, have to make sure that "edit" (which would thus be "text + enter + change") followed by a step "click" does trigger a "blur" in between :pray: Thanks :pray:

**Comment by pipu-odoo** [comment]
We'll add fill_blur in helpers to be complete in tour helpers.
However, this will be done in an other PR because to difficult to maintain this one.
